### ISSUE TITLE
Proxying fails when acting as https proxy with client curl 7.71.0, maybe TLS 1.3 problem

### ISSUE THREAD

**Issue by brianfris** [mitmproxy/mitmproxy#4154]
#### Problem Description
Using curl 7.71.0 I am unable to connect through mitmproxy acting in regular mode as an https proxy.  I'm wondering if it is related to TLS 1.3 support.  Mitmproxy either fails to connect to the destination server "Cannot connect to server, no server address given" or it hangs in the connect attempt if I disable the option "upstream_cert".

I first noticed the problem with curl 7.58.0 from Ubuntu 18.04, but the steps use the latest version of curl to try and rule out any old bugs in curl.  In my steps to reproduce I didn't bother to setup the CA certs for either proxy verificiation or the host verification, because it kept things simple and didn't affect the results as far as I could tell.  In this example it attempts to connect to www.google.com, but it seems to happen for any address I can think of.  

I'll be honest that I can't tell if it is an issue with mitmproxy or curl, but since TLS 1.3 support in mitmproxy is fairly new I thought I'd ask here first.

Thanks in advance

#### Steps to reproduce the behavior:
I took stock docker images for mitmproxy and curl to reproduce.  The first container is mitmdump and the second is a curl container using the first as an https_proxy/http_proxy.  I found the docker IP of the first container using "docker inspect", which is why in the curl command the proxy is set to 172.17.0.2.

1. Run mitmdump in first docker container.
2. Run curl command from second docker container, using mitmdump as https proxy  

```
~]$ docker run --rm curlimages/curl:latest --version
curl 7.71.1-DEV (x86_64-pc-linux-musl) libcurl/7.71.1-DEV OpenSSL/1.1.1g zlib/1.2.11 libssh2/1.9.0 nghttp2/1.40.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets

~]$ docker run --rm -it -p 8080:8080 mitmproxy/mitmproxy:latest mitmdump -p 8080 --set confdir=/home/mitmproxy --set console_eventlog_verbosity=debug --set termlog_verbosity=debug
Proxy server listening at http://*:8080
172.17.0.5:38214: clientconnect
::ffff:172.17.0.5:38214: Establish TLS with client
::ffff:172.17.0.5:38214: ALPN for client: b'http/1.1'
172.17.0.5:38214: Cannot connect to server, no server address given.
172.17.0.5:38214: ProtocolException('Cannot connect to server, no server address given.')
172.17.0.5:38214: clientdisconnect

~]$ docker run --rm curlimages/curl:latest -v -s --insecure --proxy-insecure --proxy https://172.17.0.2:8080 https://www.google.com
*   Trying 172.17.0.2:8080...
* Connected to 172.17.0.2 (172.17.0.2) port 8080 (#0)
* ALPN, offering http/1.1
* error setting certificate verify locations, continuing anyway:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [21 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [1663 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Proxy certificate:
*  subject: [NONE]
*  start date: Aug 12 19:31:06 2020 GMT
*  expire date: Aug 14 19:31:06 2021 GMT
*  issuer: CN=mitmproxy; O=mitmproxy
*  SSL certificate verify result: self signed certificate in certificate chain (19), continuing anyway.
* allocate connect buffer!
* Establish HTTP proxy tunnel to www.google.com:443
} [5 bytes data]
> CONNECT www.google.com:443 HTTP/1.1
> Host: www.google.com:443
> User-Agent: curl/7.71.1-DEV
> Proxy-Connection: Keep-Alive
>
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [233 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [233 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
< HTTP/1.1 502 Bad Gateway
< Server: mitmproxy 5.2
< Connection: close
< Content-Length: 285
< Content-Type: text/html
<
* Received HTTP code 502 from proxy after CONNECT
* CONNECT phase completed!
* Closing connection 0
} [5 bytes data]
```

3. Try again with upstream_cert disabled.  I saw in proxy/protocol/tls.py that one case that can cause it to attempt to establish the server connection is if upstream_cert is True.  So I changed that setting and tried again.  This time mitmproxy did attempt the connection to google, but it hangs for 5 minutes until it times out.  I believe the hang is in tls.py's call to do_handshake.
```
 ~]$ docker run --rm -it -p 8080:8080 mitmproxy/mitmproxy:latest mitmdump -p 8080 --set confdir=/home/mitmproxy --set console_eventlog_verbosity=debug --set termlog_verbosity=debug --set upstream_cert=false
Proxy server listening at http://*:8080
172.17.0.5:38254: clientconnect
::ffff:172.17.0.5:38254: Establish TLS with client
::ffff:172.17.0.5:38254: ALPN for client: b'http/1.1'
::ffff:172.17.0.5:38254: Set new server address: www.google.com:443
::ffff:172.17.0.5:38254: Establish TLS with client
      <mitmproxy hangs here until the connection times out>
      <it looks like in tcp.py BaseHandler.convert_to_tls it hangs at the call to do_handshake>
::ffff:172.17.0.5:38254: Failed to send error response to client: ClientHandshakeException('Cannot establish TLS with client (sni: www.google.com): TlsException("SSL handshake error: Error([(\'SSL routines\', \'ssl3_read_bytes\', \'unexpected message\')])")')
172.17.0.5:38254: CONNECT www.google.com:443
 << Cannot establish TLS with client (sni: www.google.com): TlsException("SSL handshake error: Error([('SSL routines', 'ssl3_read_bytes', 'unexpected message')])")
172.17.0.5:38254: clientdisconnect

	  
	  
 ~]$ docker run --rm curlimages/curl:latest -v -s --insecure --proxy-insecure --proxy https://172.17.0.2:8080 https://www.google.com
*   Trying 172.17.0.2:8080...
* Connected to 172.17.0.2 (172.17.0.2) port 8080 (#0)
* ALPN, offering http/1.1
* error setting certificate verify locations, continuing anyway:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [21 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [1663 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Proxy certificate:
*  subject: [NONE]
*  start date: Aug 12 19:38:04 2020 GMT
*  expire date: Aug 14 19:38:04 2021 GMT
*  issuer: CN=mitmproxy; O=mitmproxy
*  SSL certificate verify result: self signed certificate in certificate chain (19), continuing anyway.
* allocate connect buffer!
* Establish HTTP proxy tunnel to www.google.com:443
} [5 bytes data]
> CONNECT www.google.com:443 HTTP/1.1
> Host: www.google.com:443
> User-Agent: curl/7.71.1-DEV
> Proxy-Connection: Keep-Alive
>
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [233 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [233 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
< HTTP/1.1 200 Connection established
<
* Proxy replied 200 to CONNECT request
* CONNECT phase completed!
* ALPN, offering h2
* ALPN, offering http/1.1
* error setting certificate verify locations, continuing anyway:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* SSL re-using session ID
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [604 bytes data]
* CONNECT phase completed!
* CONNECT phase completed!
<<hangs here for 5 mins until timeout>>
* Operation timed out after 300498 milliseconds with 0 out of 0 bytes received
* Closing connection 0
} [5 bytes data]
* TLSv1.3 (OUT), TLS alert, close notify (256):
} [2 bytes data]
```

4.  Compare with curl 7.47.0 to show that it works when TLS 1.3 was not involved
```
~]$ docker run --rm -it -p 8080:8080 mitmproxy/mitmproxy:latest mitmdump -p 8080 --set confdir=/home/mitmproxy --set console_eventlog_verbosity=debug --set termlog_verbosity=debug --set upstream_cert=false
Proxy server listening at http://*:8080
172.17.0.3:35252: clientconnect
::ffff:172.17.0.3:35252: Set new server address: www.google.com:443
::ffff:172.17.0.3:35252: Establish TLS with client
::ffff:172.17.0.3:35252: ALPN for client: b'http/1.1'
::ffff:172.17.0.3:35252: request
  -> Request(GET /)
::ffff:172.17.0.3:35252: serverconnect
  -> ('www.google.com', 443)
::ffff:172.17.0.3:35252: Establish TLS with server
::ffff:172.17.0.3:35252: ALPN selected by server: http/1.1
::ffff:172.17.0.3:35252: response
  -> Response(200 OK, text/html; charset=ISO-8859-1, 12.5k)
172.17.0.3:35252: GET https://www.google.com/
               << 200 OK 12.5k
::ffff:172.17.0.3:35252: serverdisconnect
  -> ('www.google.com', 443)
172.17.0.3:35252: clientdisconnect

# From a docker image of ubuntu:16.04 and then installing curl via apt-get
:/# curl -V
curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets

:/# curl -v -s --insecure  --proxy https://172.17.0.2:8080 https://www.google.com
* Rebuilt URL to: https://www.google.com/
*   Trying 172.17.0.2...
* Connected to 172.17.0.2 (172.17.0.2) port 8080 (#0)
* Establish HTTP proxy tunnel to www.google.com:443
> CONNECT www.google.com:443 HTTP/1.1
> Host: www.google.com:443
> User-Agent: curl/7.47.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 Connection established
<
* Proxy replied OK to CONNECT request
* found 127 certificates in /etc/ssl/certs/ca-certificates.crt
* found 508 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: www.google.com (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=www.google.com
*        start date: Wed, 12 Aug 2020 19:48:57 GMT
*        expire date: Sat, 14 Aug 2021 19:48:57 GMT
*        issuer: CN=mitmproxy,O=mitmproxy
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET / HTTP/1.1
> Host: www.google.com
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Fri, 14 Aug 2020 19:48:57 GMT
... lots of content comes back successfully ...
```

#### System Information
 ~]$ docker run --rm -it -p 8080:8080 mitmproxy/mitmproxy:latest mitmdump --version
Mitmproxy: 5.2
Python:    3.8.2
OpenSSL:   OpenSSL 1.1.1g  21 Apr 2020
Platform:  Linux-3.10.0-862.2.3.el7.x86_64-x86_64-with

**Comment by mhils** [comment]
Thanks for the super detailed report steps, this is fantastic to work with! :cake:
So technically mitmproxy only is an HTTP proxy, not an HTTPS proxy (which is a relatively new concept that includes a second, initial client -> proxy TLS connection). We're working to add that, but TLS-over-TLS is tricky with socket APIs and requires significant changes (#1775). 

Long story short, it looks like curl recently added HTTPS proxy support and just interpreted https as http in the proxy specification before that. Does `--proxy http://172.17.0.2:8080` work instead of `--proxy https://172.17.0.2:8080`?

**Comment by brianfris** [comment]
Yup, changing the proxy to http://172.17.0.2:8080 does in fact work.  I just took for granted that curl was making an https connection to the proxy in the older versions when the url started with https://.  Trying to track down where the hanging was occurring I think I got just a glimpse into how tricky that will be.  I'll definitely keep an eye on #1775 

Thanks for the quick response.  Much appreciated!

BTW, awesome project!
